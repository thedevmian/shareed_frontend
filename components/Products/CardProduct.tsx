import { useState } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import formatMoney from "../../lib/products/formatMoney";
import QuickAddButton from "./QuickAddButton";
import AddToFavoriteButton from "./AddToFavoriteButton";
import Link from "next/link";

const ProductContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  position: relative;

  :hover {
    cursor: pointer;
    background-color: var(--main-bg-color-light);
  }
`;

const ImageContainer = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: var(--main-bg-color-light);
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
`;

const ProductImage = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

const ProductName = styled.h2`
  font-size: var(--font-size-md);
  text-transform: uppercase;
  margin: 0;
  padding-left: 1rem;
`;

const ProductPrice = styled.p`
  font-size: var(--font-size-xs);
  font-weight: bold;
  text-transform: uppercase;
  margin: 0;
  padding-right: 1rem;
`;

const ProductInfo = styled.div`
  display: flex;
  padding: 0.5rem 0;
  width: 100%;

  flex-direction: row;
  align-items: center;
  justify-content: space-between;

  background-color: var(--main-bg-color-light);
  color: var(--main-text-color);
`;
const ProductHoverInfo = styled.div`
  width: 100%;
  height: 2rem;
  display: flex;
  visibility: visible;
  flex-direction: row;
  justify-content: flex-end;
  background-color: var(--main-bg-color-light);
  color: var(--main-text-color);
`;

const CardProduct = ({ product }: any) => {
  const { id, name, price, photo } = product;
  const [hover, setHover] = useState(false);

  const handleMouseEnter = () => {
    setHover(true);
  };

  const handleMouseLeave = () => {
    setHover(false);
  };

  return (
    <ProductContainer
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <AddToFavoriteButton />
      <Link href="/product/[id]" as={`/product/${id}`}>
        <ImageContainer>
          <ProductImage
            src={photo?.[0]?.image.publicUrl || "/static/images/thumbCover.jpg"}
            alt={photo?.[0]?.altText}
          />
        </ImageContainer>
      </Link>
      <ProductInfo>
        <ProductName>{name}</ProductName>
        <ProductPrice>{formatMoney(price as number)}</ProductPrice>
      </ProductInfo>
      <ProductHoverInfo>
        {hover && <QuickAddButton productId={id} />}
      </ProductHoverInfo>
    </ProductContainer>
  );
};

CardProduct.propTypes = {
  product: PropTypes.object.isRequired,
};

export default CardProduct;
