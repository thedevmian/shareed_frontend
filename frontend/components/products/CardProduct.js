import styled from 'styled-components';
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import formatMoney from '../../lib/formatMoney';
import AddToCartButton from './AddToCartButton';
import CardAddToFavoriteButton from './CardAddToFavoriteButton';

const ProductContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  position: relative;

  :hover {
    cursor: pointer;
    background-color: var(--main-bg-color-light);
  }
`;

const ImageContainer = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: var(--main-bg-color-light);
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
`;

const ProductImage = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

const ProductName = styled.h2`
  font-size: var(--font-size-md);
  text-transform: uppercase;
  margin: 0;
  padding-left: 1rem;
`;

const ProductPrice = styled.p`
  font-size: var(--font-size-xs);
  font-weight: bold;
  text-transform: uppercase;
  margin: 0;
  padding-right: 1rem;
`;

const ProductInfo = styled.div`
  display: flex;
  padding: 0.5rem 0;
  width: 100%;

  flex-direction: row;
  align-items: center;
  justify-content: space-between;

  background-color: var(--main-bg-color-light);
  color: var(--main-text-color);
`;
const ProductHoverInfo = styled.div`
  width: 100%;
  height: 2rem;
  display: flex;
  visibility: visible;
  flex-direction: row;
  justify-content: flex-end;
  background-color: var(--main-bg-color-light);
  color: var(--main-text-color);
`;

const CardProduct = ({ product }) => {
  const { id, name, price, photo } = product;
  const [hover, setHover] = useState(false);

  const handleMouseEnter = () => {
    setHover(true);
  };

  const handleMouseLeave = () => {
    setHover(false);
  };

  console.log(hover);
  return (
    <ProductContainer
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <CardAddToFavoriteButton />
      <ImageContainer>
        <ProductImage src={photo[0]?.image.publicUrl} alt={photo[0]?.altText} />
      </ImageContainer>
      <ProductInfo>
        <ProductName>{name}</ProductName>
        <ProductPrice>{formatMoney(price)}</ProductPrice>
      </ProductInfo>
      <ProductHoverInfo>
        {hover && <AddToCartButton productId={id} />}
      </ProductHoverInfo>
    </ProductContainer>
  );
};

CardProduct.propTypes = {
  product: PropTypes.object.isRequired,
};

export default CardProduct;
