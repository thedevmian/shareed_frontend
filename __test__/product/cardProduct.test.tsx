import { render, screen } from "@testing-library/react";
import CardProduct from "components/Products/CardProduct";
import { MockedProvider } from "@apollo/react-testing";
import "@testing-library/jest-dom";

const fakeProduct = {
  id: "abc123",
  name: "Product Name",
  description: "Product Description",
  price: 100,
  user: {
    id: "abc123",
    name: "User Name",
  },
  photo: [
    {
      id: "abc123",
      image: {
        publicUrl: "image.jpg",
      },
      altText: "Product Name",
    },
  ],
};

describe("Single card product", () => {
  it("should render the product detail component", () => {
    const { container } = render(
      <MockedProvider>
        <CardProduct product={fakeProduct} />
      </MockedProvider>
    );
    expect(screen.getByText("Product Name")).toBeInTheDocument();
    expect(screen.getAllByRole("img")).toHaveLength(1);
    // photo
    const image = container.querySelector("img");
    expect(image).toHaveAttribute("src", "image.jpg");
  });

  it("should render the product detail component with no image", () => {
    const { container } = render(
      <MockedProvider>
        <CardProduct product={{ ...fakeProduct, photo: [] }} />
      </MockedProvider>
    );
    expect(screen.getByText("Product Name")).toBeInTheDocument();

    const image = container.querySelector("img");
    expect(image).toHaveAttribute("src", "/static/images/thumbCover.jpg");
  });
  it("snapshot", () => {
    const { container } = render(
      <MockedProvider>
        <CardProduct product={fakeProduct} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
