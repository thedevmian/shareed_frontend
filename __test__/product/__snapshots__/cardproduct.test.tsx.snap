// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single card product snapshot 1`] = `
<div>
  <div
    class="sc-58f62d77-0 kGfDRd"
  >
    <div
      class="sc-22a4236-0 gAxNLm"
    >
      <button
        class="sc-22a4236-1 bMLVPZ btn btn-light m-0 text-left"
        type="button"
      >
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
    <div
      class="sc-58f62d77-1 abLvM"
    >
      <img
        alt="Product Name"
        class="sc-58f62d77-2 hkyfYH"
        src="image.jpg"
      />
    </div>
    <div
      class="sc-58f62d77-5 jaqOHa"
    >
      <h2
        class="sc-58f62d77-3 dJPdHi"
      >
        Product Name
      </h2>
      <p
        class="sc-58f62d77-4 fGfVkk"
      >
        $1
      </p>
    </div>
    <div
      class="sc-58f62d77-6 hueLoB"
    />
  </div>
</div>
`;
